Git flow contents for Cloudslang

lirantal opened the pull request:
Introducing Git related flows, currently includes:
1. Git clone
2. Git checkout

orius123 commented:
This is really awesome! 
@meirwah we really should come with a way that our content can run both locally and remote and not only through ssh. WDYT?

lirantal commented:
thanks @orius123, more git content to come :-)

oritstone commented:
Thanks @lirantal for your contribution
For now, the build fails on your PR. You can see in CircleCI log (https://circleci.com/gh/CloudSlang/cloud-slang-content/1067):
![image](https://cloud.githubusercontent.com/assets/7547370/8638189/a301a748-28bb-11e5-905b-593301b7215a.png)

- As a first step, you need to fix the namespaces of the flows to match the path of the files. Since our documentation is not yet fully updated in our website, look at: https://github.com/CloudSlang/CloudSlang.github.io/blob/docs_build_tool/app/markdowns/slang.md -> **"CloudSlang Best Practices"** chapter for reference on our content best practices, which are enforced by our build tool.

- In addition, we require the content to be tested.
Look at the same link above under chapter **"CloudSlang Tests"**

lirantal commented:
Thanks @oritstone!
You're absolutely correct, I missed on the flow name and namespace when copying the initial file from another flow.

How can I run the tests locally to confirm they are working?

oritstone commented:
Look at the **"CloudSlang Build Tool"** section in: https://github.com/CloudSlang/CloudSlang.github.io/blob/docs_build_tool/app/markdowns/slang.md

meirwah commented:
@orius123 , in O they solved this by wrapping local+remote(SSH) flows with a wrapper flow, that decides (based on input) which flow to trigger in runtime.. we can do the same..

orius123 commented:
Sounds good to me. 
Lets not call it `ssh_flow` though :smile:

lirantal commented:
thanks @oritstone the build tool was quite easy to get running.

lirantal commented:
yay! now it's passing the CI :)

meirwah commented:
@Bonczidai can you please review this?

Bonczidai commented on content/io/cloudslang/git/git_checkout_branch.sl:
```diff
@@ -0,0 +1,85 @@
+  inputs:
+    - host
+    - port:
+        required: false
```
missing `port` in the description section

Bonczidai commented on content/io/cloudslang/git/git_checkout_branch.sl:
```diff
@@ -0,0 +1,85 @@
+#   Inputs:
+#       - host - hostname or IP address
+#       - username - username to connect as
+#       - password - password of user
```
if the input is not required you should add in the description section `optional` keyword e.g.:
 `#   - password - optional - password of user`

lirantal replied:
sure, in this case the password is not optional :)

Bonczidai commented on content/io/cloudslang/git/git_checkout_branch.sl:
```diff
@@ -0,0 +1,85 @@
+#       - sudo_user - true or false, whether the command should execute using sudo
+#       - git_pull_remote - if git_pull is set to true then specify the remote branch to pull from
+#       - git_branch - the git branch to checkout to
+#       - git_repository_localdir - the target directory where a git repository exists and git_branch should be checked out to
```
also, I prefer to note the default value in the description - e.g.:
`- git_repository_localdir - the target directory where a git repository exists and git_branch should be checked out to - Default: /tmp/repo.git`

Bonczidai commented:
added my comments (suggestions) - in a nutshell, excellent work :+1:

probably we want to activate the test suite ([git]) on travis so it will run automatically - @orius123 can elaborate on what's the correct way for that (since now we have a build matrix there with three suites: containers, images, docker-misc)

lirantal commented:
hey guys,
I addressed all the issues, including:
1. the flow cleans up after itself
2. the flow is added to circle-ci to run the test cases

Any ideas why the tests fail? is it relevant maybe to the build failure? (my git test case tries to connect to the docker instance to do the ssh + git operation, possible that's why it's failing)

meirwah commented:
disk space issue @orius123 ?

meirwah commented:
@Bonczidai not sure Travis is the right choice here, I think it will be better to use CircleCI, and use a docker image that has Git + SSH installed on it , @orius123  WDYT?

meirwah commented:
maybe this image : https://registry.hub.docker.com/u/nepalez/git-ssh/dockerfile/

orius123 commented:
Agree

lirantal commented:
Hey guys,

When should we expect a working solution? I would like to continue with more Git content but this PR is introducing the git directory structure and some flows that I will need to create more PRs.

One workaround I can think of incase it'll take time to workout the environment there with the build system is to remove for the time being the git test suite from the builder tool and then approve the PR so I can continue.

orius123 commented:
hi @lirantal we are currently working in fixing our matrix.
we have problems with docker 1.7.0 version changes..
can you try and add a node to the matrix of travis that will run only git tests and lets see if it helps please?

lirantal commented:
why do you think it's failing now on travis?

meirwah commented:
seems like CLEANUP FAILURE...
maybe  we need to add some prints (since we do not have yet logs collected from this Travis..)

lirantal commented:
right, although I have no idea why because it's suppose to just delete the directory it cloned the git repository to. I'm not sure if it's failing due to permission issues, or anything else.

Bonczidai commented:
Hi @lirantal ,

please note that the flow is failing because the first two tasks (git clone & checkout) execute commands using ssh (inside `dind` container), yet the third task (cleanup) executes a local command (directly on travis, not inside `dind` container) - so you need to delete the directory using ssh

e.g. look at this commit: https://github.com/CloudSlang/cloud-slang-content/commit/7dc7dda4f041660c52f861183847b57b57033955

lirantal commented:
good catch. works for me, do you want to push that commit to my PR?

lirantal commented:
closing in favor of new PR which fixes the cleanup workflow: https://github.com/CloudSlang/cloud-slang-content/pull/255
